xarray/core/dask_array_compat.py

```diff
@@ -97,6 +99,95 @@ def meta_from_array(x, ndim=None, dtype=None):
         return meta
 
 
+def _validate_pad_output_shape(input_shape, pad_width, output_shape):
+    """ Dask.array.pad with mode='reflect' does not always return the correct output_shape. """
+    isint = lambda i: isinstance(i, int)
+
+    if isint(pad_width):
+        pass
+    elif len(pad_width) == 2 and all(map(isint, pad_width)):
+        pad_width = sum(pad_width)
+    elif (
+        len(pad_width) == len(input_shape)
+        and all(map(lambda x: len(x) == 2, pad_width))
+        and all((isint(i) for p in pad_width for i in p))
+    ):
+        pad_width = np.sum(pad_width, axis=1)
+    else:
+        return  # should be impossible
+
+    if not np.array_equal(np.array(input_shape) + pad_width, output_shape):
+        raise RuntimeError(
+            "There seems to be something wrong with the shape of the output of dask.array.pad, "
+            "try upgrading Dask, use a different pad mode e.g. mode='constant' or first convert "
+            "your DataArray/Dataset to one backed by a numpy array by calling the `compute()` method."
+        )
+
+
+if LooseVersion(dask_version) >= LooseVersion("0.18.1"):
+
+    def pad(array, pad_width, mode="constant", **kwargs):
+        padded = da.pad(array, pad_width, mode=mode, **kwargs)
+        # workaround for inconsistency between numpy and dask: https://github.com/dask/dask/issues/5303
+        if mode == "mean" and issubclass(array.dtype.type, np.integer):
+            return da.round(padded).astype(array.dtype)
+        _validate_pad_output_shape(array.shape, pad_width, padded.shape)
```

Comment:
This is only needed if `mode == "reflect"`?

Reply:
Yes, I have only seen the validation gone wrong when using the `mode=="reflect"`, but in principle the validation should always hold.